tests/test_sshfsmount.cpp

```diff
@@ -188,167 +219,126 @@ struct SshfsMount : public mp::test::SftpServerTest
+    CommandVector commands = GetParam().second;
 
-    std::string output{"ubuntu"};
-    auto remaining = output.size();
-    auto channel_read = make_channel_read_return(output, remaining, invoked);
-    REPLACE(ssh_channel_read_timeout, channel_read);
-
-    EXPECT_THROW(make_sshfsmount(), std::invalid_argument);
-    EXPECT_TRUE(invoked);
+    test_command_execution(commands, target);
 }
 
-TEST_F(SshfsMount, throws_when_unable_to_obtain_gid)
+TEST_P(SshfsMountExecuteThrow, test_invalid_arg_when_executing)
 {
-    bool uid_invoked{false};
-    bool gid_invoked{false};
-    auto request_exec = [this, &uid_invoked, &gid_invoked](ssh_channel, const char* raw_cmd) {
-        std::string cmd{raw_cmd};
-        if (cmd.find("id -u") != std::string::npos)
-        {
-            uid_invoked = true;
-        }
-        else if (cmd.find("id -g") != std::string::npos)
-        {
-            uid_invoked = false;
-            gid_invoked = true;
-            exit_status_mock.return_exit_code(SSH_ERROR);
-        }
-        return SSH_OK;
-    };
-    REPLACE(ssh_channel_request_exec, request_exec);
-
-    std::string output{"1000"};
-    auto remaining = output.size();
-    auto channel_read = make_channel_read_return(output, remaining, uid_invoked);
-    REPLACE(ssh_channel_read_timeout, channel_read);
-
-    EXPECT_THROW(make_sshfsmount(), std::runtime_error);
-    EXPECT_TRUE(gid_invoked);
+    MP_ASSERT_THROW_THAT(test_command_execution(GetParam().first), std::exception,
+                         Property(&std::exception::what, ContainsRegex(GetParam().second)));
 }
 
-TEST_F(SshfsMount, throws_when_gid_is_not_an_integer)
+//
+// Instantiate the parameterized tests suites from above.
+//
+
+INSTANTIATE_TEST_SUITE_P(SshfsMountThrowWhenError, SshfsMountFail,
+                         testing::Values("mkdir", "chown", "id -u", "id -g", "cd", "pwd"));
+
+// Commands to check that a version of FUSE smaller that 3 gives a correct answer.
+CommandVector fuse_cmds = {{"sudo env LD_LIBRARY_PATH=/foo/bar /baz/bin/sshfs -V", "FUSE library version: 2.9.0\n"},
+                           {"sudo env LD_LIBRARY_PATH=/foo/bar /baz/bin/sshfs -o slave -o transform_symlinks -o "
+                            "allow_other -o nonempty :\"source\" \"target\"",
+                            "don't care\n"}};
+
+// Commands to check that the server correctly executes commands.
+CommandVector exec_cmds = {
+    {"sudo /bin/bash -c 'P=\"/home/ubuntu/target\"; while [ ! -d \"$P/\" ]; do P=${P%/*}; done; echo $P/'",
+     "/home/ubuntu/\n"},
+    {"sudo /bin/bash -c 'cd \"/home/ubuntu/\" && mkdir -p \"target\"'", "\n"},
+    {"sudo /bin/bash -c 'cd \"/home/ubuntu/\" && chown -R 1000:1000 target'", "\n"}};
+
+// Commands to check that the server works if an absolute path is given.
+CommandVector absolute_cmds = {
+    {"sudo /bin/bash -c 'P=\"/home/ubuntu/target\"; while [ ! -d \"$P/\" ]; do P=${P%/*}; done; echo $P/'",
+     "/home/ubuntu/\n"}};
```

Comment:
Here, too, I think like we're missing a couple checks. We should verify what happens on different return values of this command:
1. full path already existing
2. none of the path existing

Ideally we'd check that neither `mkdir` or `chown` are executed unless necessary. If you made the change to add an optional `fail_cmd` I suggested above, you could `ASSERT_NO_THROW` with a `CommandVector{…}` returning the value of `target` and failing on either `mkdir` or `chown`.

Reply:
That is a neat idea. In fact, I am just discovering that, when no directory must be created, `mkdir "."` and `chown -R 1000:1000 "."` are executed. At least in an old version of the code. That should be corrected when merging the `~` handling.